File: Perl/2024/2024-001/part-002/main.pl
use strict;
use warnings;

# Define the input file path
my $file_path = "input.txt";

# Initialize arrays for the two lists
my (@left_list, @right_list);

# Read the file and extract the two columns
open my $fh, '<', $file_path or die "Could not open file '$file_path': $!";
while (my $line = <$fh>) {
    chomp $line;
    my @parts = split /\s+/, $line;
    if (@parts == 2) {
        push @left_list,  $parts[0];
        push @right_list, $parts[1];
    }
}
close $fh;

# Create a frequency hash (dictionary) for the right list
my %right_list_counts;
foreach my $num (@right_list) {
    $right_list_counts{$num}++;
}

# Calculate the similarity score
my $similarity_score = 0;
foreach my $left (@left_list) {
    my $count = $right_list_counts{$left} // 0;  # Default to 0 if not found
    $similarity_score += $left * $count;
}

# Print the result
print "Similarity Score: $similarity_score\n";

File: PHP/2024/2024-001/part-002/main.php
<?php
// Define the file path
$filePath = 'input.txt';

// Check if the file exists
if (!file_exists($filePath)) {
    die("File not found: " . $filePath);
}

// Load the file content into an array
$fileContent = file($filePath, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);

// Initialize the left and right lists
$leftList = [];
$rightList = [];

// Parse the file content
foreach ($fileContent as $line) {
    $columns = preg_split('/\s+/', trim($line));
    if (count($columns) >= 2) {
        $leftList[] = (int)$columns[0];
        $rightList[] = (int)$columns[1];
    }
}

// Count occurrences in the right list
$rightListCounts = array_count_values($rightList);

// Calculate the similarity score
$similarityScore = 0;
foreach ($leftList as $left) {
    $similarityScore += $left * ($rightListCounts[$left] ?? 0);
}

// Output the result
echo "Similarity Score: " . $similarityScore;
?>
